### PR TITLE
Kubernetes Helm chart

### PR DESCRIPTION
**Pull request by zakkg3**
Signed-off-by: Nico <zakkg3@gmail.com>

### CODE REVIEW

File: syslog-ng-helm-chart/templates/tests/test-connection.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  labels:
+    {{- include "syslog-ng.labels" . | nindent 4 }}
+  annotations:
+    "helm.sh/hook": test-success
+spec:
+  containers:
+    - name: wget
+      image: busybox
+      command: ['wget']
+      args: ['{{ include "syslog-ng.fullname" . }}:{{ .Values.service.port }}']
```
Comment (bazsi) [inline syslog-ng-helm-chart/templates/tests/test-connection.yaml:14]: Is there anything that listens on http within the container? I couldn't find anything.

File: syslog-ng-helm-chart/.helmignore [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+*.swp
+*.bak
+*.tmp
+*.orig
+*~
+# Various IDEs
+.project
+.idea/
+*.tmproj
+.vscode/
```
Comment (bazsi) [inline syslog-ng-helm-chart/.helmignore:23]: Can you remove the irrelevant files from here?

File: syslog-ng-helm-chart/templates/NOTES.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get --namespace {{ .Release.Namespace }} svc -w {{ include "syslog-ng.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ include "syslog-ng.fullname" . }} --template "{{"{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}"}}")
+  echo http://$SERVICE_IP:{{ .Values.service.port }}
+{{- else if contains "ClusterIP" .Values.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app.kubernetes.io/name={{ include "syslog-ng.name" . }},app.kubernetes.io/instance={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
```
Comment (bazsi) [inline syslog-ng-helm-chart/templates/NOTES.txt]: Again, I think nothing listens on 127.0.0.1 port 8080

File: syslog-ng-helm-chart/templates/compress-job.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+metadata:
+  name: job-script
+data:
+  entrypoint.sh: |-
+      #!/bin/bash
+      ls -lh /var/log/*-syslog
+      for filename in /var/log/*-syslog; do
+       if [[ $(find "$filename" -mtime +{{ .Values.compressor.retention_days }} -print) ]]; then
+         echo "$filename is older than {{ .Values.compressor.retention_days }} days, Compressing."
+         tar -zcvf "$filename"-$(date +%d-%m-%y).tar.gz $filename
```
Comment (bazsi) [inline syslog-ng-helm-chart/templates/compress-job.yaml:14]: Usually logfiles are merely compressed via gzip and not encapsulated in a tar envelope. Is there a reason it is better this way when running on kubernetes?

File: syslog-ng-helm-chart/templates/compress-job.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: job-script
+data:
+  entrypoint.sh: |-
+      #!/bin/bash
+      ls -lh /var/log/*-syslog
+      for filename in /var/log/*-syslog; do
+       if [[ $(find "$filename" -mtime +{{ .Values.compressor.retention_days }} -print) ]]; then
```
Comment (bazsi) [inline syslog-ng-helm-chart/templates/compress-job.yaml:12]: I think there should be two stages of retention,

1) data should be compressed after a day
2) compressed files should be removed after the retention period

File: syslog-ng-helm-chart/templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  ###
+        {{- if .Values.logrotate.enable }}
+          - name: logrotate
+            securityContext:
+              {{- toYaml .Values.securityContext | nindent 12 }}
+            image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+            imagePullPolicy: {{ .Values.image.pullPolicy }}
+            command:
+              -"/bin/logrotate"
+              -"--help"
```
Comment (bazsi) [inline syslog-ng-helm-chart/templates/deployment.yaml]: I not a kubernetes expert, but I don't see why "logrotate --help" makes sense here. When is it executed?

File: syslog-ng-helm-chart/values.yaml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    # keep hostnames from source host
+    keep_hostname(yes);
+
+    # use ISO8601 timestamps
+    ts_format(iso);
+  };
+  log {
+  	source {
+  		network();
```
Comment (bazsi) [inline syslog-ng-helm-chart/values.yaml:29]: Could we perhaps use default-network-drivers() here?

That would enable automatic parsing of a number of sources using the app-parser() framework. It also opens up a few more ports (e.g. tls enabled stuff).

Tls support would be useful, e.g. the management of the x.509 certificates.

File: syslog-ng-helm-chart/values.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  	};
+  	destination { file("/var/log/${YEAR}-${MONTH}-syslog"); };
+  };
+
+storage:
+  enable: True
+  storageClass: netapp
```
Comment (bazsi) [inline syslog-ng-helm-chart/values.yaml]: What is "netapp" here? Is it a setting that makes sense for a broad set of users or it's local to you?

File: syslog-ng-helm-chart/values.yaml [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    keep_hostname(yes);
+
+    # use ISO8601 timestamps
+    ts_format(iso);
+  };
+  log {
+  	source {
+  		network();
+  	};
+  	destination { file("/var/log/${YEAR}-${MONTH}-syslog"); };
```
Comment (bazsi) [inline syslog-ng-helm-chart/values.yaml:31]: Monthly rotation might be too rare for a default.

File: syslog-ng-helm-chart/values.yaml [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+podSecurityContext: {}
+  # fsGroup: 2000
+
+securityContext: {}
+  # capabilities:
+  #   drop:
+  #   - ALL
+  # readOnlyRootFilesystem: true
+  # runAsNonRoot: true
+  # runAsUser: 1000
```
Comment (bazsi) [inline syslog-ng-helm-chart/values.yaml:73]: Syslog-ng manages its own capabilities as well, just make sure the two doesn't collide.

File: syslog-ng-helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/templates/deployment.yaml]: Thats a mistake. i was testing a co-container in the same pod, for logrotate but finally decide to use a cronjob, this was never included on the deployment as in values there is no logrotate.enable. I will delete this and commit :D

File: syslog-ng-helm-chart/.helmignore (hunk 2 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/.helmignore:23]: this is a helm best practice, helm create thats for you when you create a new chart. Its based on the community experience. This is not relative to my workspace. But let me know if you still want to get rid of this. makes no difference ;)

File: syslog-ng-helm-chart/templates/tests/test-connection.yaml (hunk 1 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/templates/tests/test-connection.yaml:14]: This is helm test. It test if a container can get (using the k8s dns) something on the service.
For this you need to understand how internal dns works. It has nothing to do with syslog-ng .
https://helm.sh/docs/topics/chart_tests/

File: syslog-ng-helm-chart/templates/NOTES.txt (hunk 3 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/templates/NOTES.txt]: Yap thats the default messge from helm I had refactor the mssege as this is not a web app. 
fixed :)

File: syslog-ng-helm-chart/values.yaml (hunk 7 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/values.yaml:29]: Tls support requieres another iteration on this chart, maybe next version of this. I suggest to include this as it is now, and let the community start pushing code for all the features are missing here.
please let me know how do you like the default to config and i will include.

File: syslog-ng-helm-chart/templates/compress-job.yaml (hunk 5 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/templates/compress-job.yaml:12]: I agree on this idea and i will implement it but for now this requieres another iteration on this chart.

File: syslog-ng-helm-chart/values.yaml (hunk 8 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/values.yaml]: Ups! that was a local thing to me. i left it blank so k8s will pick up the default sotrageclass.

File: syslog-ng-helm-chart/values.yaml (hunk 9 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/values.yaml:31]: as i mentioned before, i am happy to include a default config you like. Please share it with me :)

File: syslog-ng-helm-chart/values.yaml (hunk 10 above)
Comment (zakkg3) [reply syslog-ng-helm-chart/values.yaml:73]: for now i will include your comment on the values file. 
let me know if you want to refactor this, and how, or if you prefer to remove. This is again a helm premade feature, iam not using.

**Comment by zakkg3** [comment]
Hi Just "resolved" all the comments. but iam open to change things.
 tl;dr please share with me the default config you like to include.

And for the features are still missing here. Like:

* Logrotate
* TLS
* HealthChecks
* prometheus exporter

and so on.
I suggest we include the chart as an initial version and iterate on this again a few times.
I will probably submit another pr for the compressor cronjob to include your suggestions. But i cant do it now. I have tons of work :/
thanks for the feedback!

**Comment by szemere** [comment]
Hello @zakkg3 ,
 sorry for the late delay, I try to warm this topic up. I installed **minikube** to test your scripts. There are some typos, nothing serious so far. *
 You have marked all of the comments from @bazsi as resolved. However I think there is room for further discussions in some cases. Please let me know if you are willing to to work on this PR, or as you stated: "I suggest we include the chart as an initial version and iterate on this again a few times.".
 My preference is to resolve open questions in "version 1", but If this will block your workflow I think we can go along with multiple iterations instead.


\* I have found several internal variables and multiple assumptions/branching in the scripts (along with some unused/undefined variables). Currently I am not sure if those are "helm best practices" or your unique implementation details. I might have several question in the topic. I prefer to give a more general solution.

Br,
Laci

File: syslog-ng-helm-chart/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v2
```
Comment (szemere) [inline syslog-ng-helm-chart/Chart.yaml:1]: I have helm v2.16.9 but it gives me an error message on the "apiVersion: v2" option. (Other files in this PR uses apiVersion v1)
Can you help me with some "helm install" command examples?

File: syslog-ng-helm-chart/templates/NOTES.txt [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get --namespace {{ .Release.Namespace }} svc -w {{ include "syslog-ng.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ include "syslog-ng.fullname" . }} --template "{{"{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}"}}")
+  echo http://$SERVICE_IP:{{ .Values.service.port }}
+{{- else if contains "ClusterIP" .Values.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app.kubernetes.io/name={{ include "syslog-ng.name" . }},app.kubernetes.io/instance={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Syslogng watch on tcp  601 and upd 514 por"
```
Comment (szemere) [inline syslog-ng-helm-chart/templates/NOTES.txt:14]:
```suggestion
  echo "Syslogng watch on tcp 601 and upd 514 port"
```

File: syslog-ng-helm-chart/templates/compress-job.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
+{{- if .Values.compressor.enabled }}
```
Comment (szemere) [inline syslog-ng-helm-chart/templates/compress-job.yaml:1]: We can shorter the scripts, if we eliminate the "enabled" flag, and test on **{{- if .Values.compressor }}** instead. This way the whole compressor section would be optional in the values file.

**Comment by szemere** [comment]
Hello @zakkg3 ,
 do you plan to work on this integration in the future? I would rather not let this code fell on the ground. Please let me know if I can help you.

Best regards,
Laci

**Comment by gaborznagy** [comment]
@zakkg3 We've added some changes onto yours and created a new PR out of it. See #63.

I'm closing this will be continued in #63
